### PR TITLE
Script to reclaim leftover nova servers.

### PR DESCRIPTION
**Pull request by burton-swan**
## Description of change
As requested by card:
https://canonical.leankit.com/Boards/View/115003549/129241960

To reclaim the leftover nova servers.

## QA steps
$ nova list
$ ./clean_prodstack.py
$ nova list

Check output from the script as well as nova list, leftover servers should be gone.

## Documentation changes
N/A

## Bug reference
N/A

### CODE REVIEW

File: acceptancetests/clean_prodstack.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+import re
+import os
+import sys
+import logging
+import subprocess
+
+def get_current_controller_and_models():
+    lines = subprocess.check_output(
+        ['juju', 'models'], stderr=subprocess.STDOUT).split('\n')
```
Comment (Veebers) [inline acceptancetests/clean_prodstack.py]: Use "-- format yaml" or "--format json"  and parse that into a datastructure, that you can then get the required details more easily.

For instance:
    model_output = json.loads(subprocess.check_output(['juju', 'models', '--format', 'json']))
    model_list = [m['name'] for m in  all_models['models']]
    controller_output = json.loads(subprocess.check_output(['juju', 'controllers', '--format', 'json']))
    controller_name = controller_output['current-controller']

File: acceptancetests/clean_prodstack.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        ctrl_name,
+        '\n'.join(model_list)))
+    return (ctrl_name, model_list)
+
+
+def get_reserved_machine_ipaddress():
+    role = 'admin'
+    ip_pool = []
+    ctrl_name, model_list = get_current_controller_and_models()
+    for item in model_list:
```
Comment (Veebers) [inline acceptancetests/clean_prodstack.py]: model is a better variable name then item, model_name is even better as it describes what it is.

File: acceptancetests/clean_prodstack.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        model = '{}:{}/{}'.format(ctrl_name, role, item)
+        lines = subprocess.check_output(
+            ['juju', 'show-machine', '--model', model],
+            stderr=subprocess.STDOUT)
```
Comment (Veebers) [inline acceptancetests/clean_prodstack.py]: As above, use --format yaml or json and there is no string comparison or regex required.

ip_pool.extend(
    *[machine_output['machines'][machine_num]['ip-addresses]
    for machine_num in machines_output['machines'].keys()])

File: acceptancetests/clean_prodstack.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    return ip_pool
+
+
+def get_target_server_list():
+    ip_pool = get_reserved_machine_ipaddress()
+    egrep_out = []
+    reserved_nova_server_dict = {}
+    full_nova_server_dict = {}
+    target_nova_server_dict = {}
+    temp_file = '/tmp/burton-test-nova-list.txt'
```
Comment (Veebers) [inline acceptancetests/clean_prodstack.py]: Lets use NamedTemporaryFile for temp files e.g. 

with NamedTemporaryFile() as tmp_file:
    subprocess.check_call([...], stdout=tmp_file)
    tmp_file.seek(0)  # Need to jump back to start of file after writing.
    for line in tmp_file.readlines():
        ....

File: acceptancetests/clean_prodstack.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    with open(temp_file, 'w') as f:
+        subprocess.check_call(['nova', 'list'], stdout=f)
+    if not os.path.isfile(temp_file):
```
Comment (Veebers) [inline acceptancetests/clean_prodstack.py]: A better check would be to make sure the call to nova list returned non-zero.

File: acceptancetests/clean_prodstack.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        logging.info('Failed to get nova server list, exit.')
+        sys.exit(1)
+    with open(temp_file, 'r') as stream:
+        # using readlines() as the outout from nova list is tiny
+        for line in stream.readlines():
+            if 'juju-' in line:
+                full_nova_server_dict[
```
Comment (Veebers) [inline acceptancetests/clean_prodstack.py]: This needs a comment as to what it's doing, it's not obvious what output it's trying to coerce into what format.

I *think* using "nova show --minimal" should simplify this a little (just shows uuid and name)

File: acceptancetests/clean_prodstack.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                    filter(None, line.split('|'))[0].strip()] = filter(
+                    None, line.split('|'))[1].strip()
+    if not ip_pool:
```
Comment (Veebers) [inline acceptancetests/clean_prodstack.py]: This conditional needs a comment to describe why it's doing what it's doing.

File: acceptancetests/clean_prodstack.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    for key in target_nova_server_dict:
+        try:
+            subprocess.check_output(
+                ['nova', 'delete', target_nova_server_dict[key]],
+                stderr=subprocess.STDOUT)
+            logging.info(
+                'nova server {} has been deleted.'.format(
+                target_nova_server_dict[key]))
+        except subprocess.CalledProcessError:
+            logging.info(
```
Comment (Veebers) [inline acceptancetests/clean_prodstack.py]: This logging level should surely be warning or error as an error has occurred?

**Comment by burton-swan** [comment]
$$merge$$

**Pull request merged**